File: apps/desktop/ui/src/ipc.ts
// SPDX-License-Identifier: GPL-3.0-or-later
import { invoke } from "@tauri-apps/api/core";
import { save as dialogSave, open as dialogOpen } from "@tauri-apps/plugin-dialog";

// ponytail: loose types for delta/snapshot payloads until Rust shape mirrors TS
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Args = Record<string, any>;

export async function newDoc() {
  return invoke("new_doc", {});
}

export async function snapshot() {
  return invoke("snapshot", {});
}

export async function commitTransform(args: Args) {
  return invoke("commit_transform", args);
}

export async function addPrimitive(args: Args) {
  return invoke("add_primitive", args);
}

export async function booleanOp(args: Args) {
  return invoke("boolean_op", args);
}

export async function addText(args: Args) {
  return invoke("add_text", args);
}

export async function deleteNodes(args: Args) {
  return invoke("delete", args);
}

export async function reorder(args: Args) {
  return invoke("reorder", args);
}

export async function undo() {
  return invoke("undo", {});
}

export async function redo() {
  return invoke("redo", {});
}

export async function importSvg(args: Args) {
  return invoke("import_svg", args);
}

export async function saveProject(args: Args) {
  return invoke("save_project", args);
}

export async function loadProject(args: Args) {
  return invoke("load_project", args);
}

export async function setMachine(args: Args) {
  return invoke("set_machine", args);
}

export async function listMachines() {
  return invoke("list_machines", {});
}

// --- device / cut / preset wire types (mirror driver-core::manager + desktop::device) ---

export type TransportKind =
  | { Usb: { locator: string } }
  | { Serial: { path: string; baud: number } };

export type DeviceInfo = {
  instance_id: string;
  machine_id: string;
  transport: TransportKind;
  candidate: boolean;
  // null means this cutter is attached to this computer. A Cut Host's cutters carry the id of
  // the host that owns them, which is what every call routes on.
  host: string | null;
};

export type PairedHostView = {
  id: string;
  name: string;
  address: string;
  /** Why this host cannot be reached, or null when it can. */
  unreachable: string | null;
};

export type DeviceError =
  | "Disconnected"
  | "Busy"
  | "Timeout"
  | "WriteZero"
  | { Io: string };

export type Phase =
  | "Disconnected"
  | "Connecting"
  | "Disconnecting"
  | "Idle"
  | "Sending"
  | "AwaitingConfirmation"
  | "AwaitingColorSwap"
  | "Cancelling"
  | "Failed";

/** Mirrors driver_core::CutStatus. The phase says what is happening now; `ended`
 *  says how the last job finished, which no phase can — a finished cut and a
 *  cancelled one both rest on "Idle". Actions say which buttons are legal.
 *  Nothing here needs interpreting, and nothing needs remembering. */
export type CutStatus = {
  phase: Phase;
  ended: "Completed" | "Cancelled" | null;
  actions: { cut: boolean; cancel: boolean; resume: boolean; confirm: boolean };
  pass: { index: number; total: number } | null;
  sent: { sent: number; total: number } | null;
  error: DeviceError | null;
};

/** Mirrors CutStatus::disconnected() — what to show before the first status arrives. */
export const DISCONNECTED_STATUS: CutStatus = {
  phase: "Disconnected",
  ended: null,
  actions: { cut: false, cancel: false, resume: false, confirm: false },
  pass: null,
  sent: null,
  error: null,
};

// `StateChanged` carries no payload: the event's own `status` is what changed.
export type DeviceEventKind =
  | "StateChanged"
  | { Progress: { pass_index: number; submitted_bytes: number; total_bytes: number } }
  | { PassComplete: number }
  | "JobComplete"
  | { Failed: DeviceError };

export type DeviceEvent = { job_id: number; kind: DeviceEventKind; status: CutStatus };

export type PlanCutPassSummary = {
  color: number | null;
  shape_count: number;
  node_ids: number[];
  /** Each shape's first world-space point, parallel to node_ids — where the blade lands.
   *  null is a shape whose outline flattened to nothing. */
  starts: ([number, number] | null)[];
};

export type PlanCutResponse = {
  passes: PlanCutPassSummary[];
  skipped_no_stroke: number;
  doc_revision: string;
  travel: [number, number, number, number][];
};

export type IpcError = { code: string; message: string };

// Real IpcError-derived commands reject with the serialized {code,message}
// object itself (not a string) — String(e) on those yields "[object Object]".
// Older doc-editing commands still reject with a plain string. Handle both.
export function ipcErrorMessage(e: unknown): string {
  if (e && typeof e === "object" && "message" in e) return String((e as { message: unknown }).message);
  return String(e);
}

export function ipcErrorCode(e: unknown): string | null {
  if (e && typeof e === "object" && "code" in e) return String((e as { code: unknown }).code);
  return null;
}

export async function listDevices(): Promise<DeviceInfo[]> {
  return invoke("list_devices", {});
}

export async function connectDevice(info: DeviceInfo): Promise<void> {
  return invoke("connect_device", { info });
}

export async function disconnectDevice(): Promise<void> {
  return invoke("disconnect_device", {});
}

/** Re-opens the aimed cutter's transport, locally or on its Cut Host. What clears a cancel whose
 *  stop nothing confirmed — there is no verb for declaring one confirmed. */
export async function reconnectDevice(): Promise<void> {
  return invoke("reconnect_device", {});
}

export async function getDeviceState(): Promise<CutStatus> {
  return invoke("get_device_state", {});
}

export async function getConnectedDevice(): Promise<DeviceInfo | null> {
  return invoke("get_connected_device", {});
}

export async function forceQuit(): Promise<void> {
  return invoke("force_quit", {});
}

export async function planCut(): Promise<PlanCutResponse> {
  return invoke("plan_cut", {});
}

/** A pass as the dialog has it configured: where it sits in the order, and whether it is cut. */
export type TravelPass = { color: number | null; enabled: boolean };

/** Travel replanned by the backend for the dialog's current pass list. Every planned pass
 *  must be named (disabled ones included — they are dropped from the travel, not from the
 *  list). Rejects with code "stale_plan" when the document has changed since `docRevision`
 *  was planned. */
export async function travelForOrder(
  docRevision: string,
  passes: TravelPass[],
): Promise<[number, number, number, number][]> {
  return invoke("travel_for_order", { docRevision, passes });
}

/** What a press of Cut did. `duplicate` is the Cut Host saying it had already accepted this
 *  dispatch and started nothing — the one fact the desktop cannot work out for itself, and the
 *  difference between a cutter about to move and one that never will. */
export type CutStarted = { job_id: number; duplicate: boolean };

export async function cut(request: Args): Promise<CutStarted> {
  return invoke("cut", { request });
}

export async function cancelCut(): Promise<void> {
  return invoke("cancel_cut", {});
}

export async function resumeCut(): Promise<void> {
  return invoke("resume_cut", {});
}

export async function confirmPassDone(): Promise<void> {
  return invoke("confirm_pass_done", {});
}

export async function listHosts(): Promise<PairedHostView[]> {
  return invoke("list_hosts", {});
}

/** The fingerprint a host presents, for the operator to confirm. Sends no token: it runs
 *  before the operator has confirmed the host's identity. */
export async function probeHost(address: string): Promise<string> {
  return invoke("probe_host", { address });
}

export async function testHost(address: string, token: string, fingerprint: string): Promise<DeviceInfo[]> {
  return invoke("test_host", { address, token, fingerprint });
}

/** A Cut Host already paired at this address, if there is one. `sameFingerprint: false` means the
 *  host's certificate changed since it was paired — a reinstall, or something worth worrying
 *  about, and the operator is the only one who knows which. */
export type ExistingPairing = { id: string; name: string; sameFingerprint: boolean };

export async function existingPairing(address: string, fingerprint: string): Promise<ExistingPairing | null> {
  return invoke("existing_pairing", { address, fingerprint });
}

export async function pairHost(name: string, address: string, token: string, fingerprint: string): Promise<PairedHostView> {
  return invoke("pair_host", { name, address, token, fingerprint });
}

// `force` is the operator accepting that a host which cannot be asked may still be cutting. Only
// offered once an unforced attempt has been refused — see `forgetFrom`.
export async function forgetHost(id: string, force: boolean): Promise<void> {
  return invoke("forget_host", { id, force });
}

export async function listPresets(machineId: string) {
  return invoke("list_presets", { machineId });
}

export async function machineCaps(machineId: string) {
  return invoke("machine_caps", { machineId });
}

export async function savePreset(p: Args) {
  return invoke("save_preset", { p });
}

export async function deletePreset(id: string) {
  return invoke("delete_preset", { id });
}

const CUT_FILTER = [{ name: "cuthulhu project", extensions: ["cut"] }];

export async function pickSavePath(): Promise<string | null> {
  return dialogSave({ defaultPath: "cuthulhu-project.cut", filters: CUT_FILTER });
}

export async function pickOpenPath(): Promise<string | null> {
  const r = await dialogOpen({ multiple: false, filters: CUT_FILTER });
  return typeof r === "string" ? r : null;
}

// --- trace wire types ---

export type TraceControlsDto = {
  mode: "binary" | "color";
  speckle: number;
  smoothing: number;
  detail: number;
  colors: number;
};
// Mirrors trace::ControlSpec. No range, default, or step is written on this side — the whole point
// of the command below is that these numbers have one home.
export type ControlSpec = {
  name: "speckle" | "smoothing" | "detail" | "colors";
  label: string;
  help: string;
  min: number;
  max: number;
  step: number;
  default: number;
  colorOnly: boolean;
};
export type TraceControlSpecsDto = {
  controls: ControlSpec[];
  defaultMode: "binary" | "color";
  maxDim: number;
};
export type TraceResultDto = { svg: string; pathCount: number; widthPx: number; heightPx: number; downscaled: boolean };

export async function traceControls(): Promise<TraceControlSpecsDto> {
  return invoke("trace_controls", {});
}
// Sorted installed family names; empty on a system with no fonts (a state, not an error).
export async function listFonts(): Promise<string[]> {
  return invoke("list_fonts", {});
}
export async function traceImage(args: { path: string; controls: TraceControlsDto }): Promise<TraceResultDto> {
  return invoke("trace_image", args);
}
export async function loadImagePreview(args: { path: string }): Promise<string> {
  return invoke("load_image_preview", args);
}
// Goes through Rust rather than the dialog plugin directly: the backend records what the user
// picked and refuses to trace anything else, so choosing the file here is what grants access.
export async function pickImagePath(): Promise<string | null> {
  return invoke("pick_image", {});
}

File: apps/desktop/src/ipc.rs
// SPDX-License-Identifier: GPL-3.0-or-later
use std::path::PathBuf;
use std::sync::Mutex;
use document::{CutLineType, Delta, MachineProfile, NodeId, PresetAssignment, ShapeKind};
use driver_core::{CutStatus, DeviceInfo, HostId, MachineCaps};
use geometry::{Affine, BoolOp};
use crate::device::{plan_cut_response, CutRequest, CutStarted, DeviceManagerHandle, ExistingPairing, IpcError, PairedHostView, PlanCutResponse, TravelPassDto};
use crate::state::AppState;
use cutplan::presets::MaterialPreset;
use cutplan::Grouping;

pub type AppStateHandle = Mutex<AppState>;

#[tauri::command]
pub fn new_doc(state: tauri::State<AppStateHandle>) -> Result<String, String> {
    Ok(state.lock().unwrap().new_doc())
}

#[tauri::command]
pub fn snapshot(state: tauri::State<AppStateHandle>) -> Result<String, String> {
    Ok(state.lock().unwrap().snapshot())
}

#[tauri::command]
pub fn commit_transform(state: tauri::State<AppStateHandle>, ids: Vec<NodeId>, m: Affine) -> Result<Delta, String> {
    state.lock().unwrap().commit_transform(ids, m).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn add_primitive(state: tauri::State<AppStateHandle>, parent: NodeId, kind: ShapeKind) -> Result<Delta, String> {
    state.lock().unwrap().add_primitive(parent, kind).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn boolean_op(state: tauri::State<AppStateHandle>, ids: Vec<NodeId>, op: BoolOp) -> Result<Delta, String> {
    state.lock().unwrap().boolean_op(ids, op).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn add_text(state: tauri::State<AppStateHandle>, parent: NodeId, family: String, size_mm: f64, text: String) -> Result<Delta, String> {
    state.lock().unwrap().add_text(parent, family, size_mm, text).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn delete(state: tauri::State<AppStateHandle>, ids: Vec<NodeId>) -> Result<Delta, String> {
    state.lock().unwrap().delete(ids).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn reorder(state: tauri::State<AppStateHandle>, id: NodeId, new_index: usize) -> Result<Delta, String> {
    state.lock().unwrap().reorder(id, new_index).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn set_cut_line_type(state: tauri::State<AppStateHandle>, ids: Vec<NodeId>, value: CutLineType)
    -> Result<Delta, String> {
    state.lock().unwrap().set_cut_line_type(ids, value).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn set_material_preset(state: tauri::State<AppStateHandle>, ids: Vec<NodeId>, value: PresetAssignment)
    -> Result<Delta, String> {
    state.lock().unwrap().set_material_preset(ids, value).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn undo(state: tauri::State<AppStateHandle>) -> Result<Option<Delta>, String> {
    Ok(state.lock().unwrap().undo())
}

#[tauri::command]
pub fn redo(state: tauri::State<AppStateHandle>) -> Result<Option<Delta>, String> {
    Ok(state.lock().unwrap().redo())
}

#[tauri::command]
pub fn import_svg(state: tauri::State<AppStateHandle>, bytes: Vec<u8>, parent: NodeId) -> Result<(Delta, Vec<String>), String> {
    state.lock().unwrap().import_svg(bytes, parent).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn save_project(state: tauri::State<AppStateHandle>, path: PathBuf) -> Result<(), String> {
    state.lock().unwrap().save_project(&path).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn load_project(state: tauri::State<AppStateHandle>, path: PathBuf) -> Result<String, String> {
    state.lock().unwrap().load_project(&path).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn set_machine(state: tauri::State<AppStateHandle>, machine_id: String) -> Result<(), String> {
    state.lock().unwrap().set_machine(&machine_id).map_err(|e| format!("{e:?}"))
}

#[tauri::command]
pub fn list_machines(state: tauri::State<AppStateHandle>) -> Result<Vec<MachineProfile>, String> {
    Ok(state.lock().unwrap().list_machines())
}

// --- device / cut / preset commands: operate over DeviceManagerHandle, never AppStateHandle's mutex ---

// async: dials every paired host to refresh its device list (30s budget per host, by design —
// see DeviceManagerHandle::list_devices) — on the main thread one wedged Pi freezes the window.
#[tauri::command(async)]
pub fn list_devices(dev: tauri::State<DeviceManagerHandle>) -> Result<Vec<DeviceInfo>, IpcError> {
    Ok(dev.list_devices())
}

// async: worker may be busy mid-transmit, and USB open has real latency.
#[tauri::command(async)]
pub fn connect_device(dev: tauri::State<DeviceManagerHandle>, info: DeviceInfo) -> Result<(), IpcError> {
    dev.connect(info)
}

#[tauri::command(async)]
pub fn disconnect_device(dev: tauri::State<DeviceManagerHandle>) -> Result<(), IpcError> {
    dev.disconnect()
}

// async for the same reasons as connect_device, plus a remote arm that talks to the Pi.
#[tauri::command(async)]
pub fn reconnect_device(dev: tauri::State<DeviceManagerHandle>) -> Result<(), IpcError> {
    dev.reconnect()
}

// async: non-blocking for a local device — the status is published rather than asked of the
// worker (see DeviceManagerHandle::status) — but for a remote one this polls the host over the
// network, bounded by roughly 2x STATUS_POLL_TIMEOUT (reconnect leg, then body-read leg) with
// the DNS/mDNS resolve bounded inside the first leg's deadline, and it is called after every
// connect/cancel/resume/confirm, so it must not run on the main thread.
#[tauri::command(async)]
pub fn get_device_state(dev: tauri::State<DeviceManagerHandle>) -> Result<CutStatus, IpcError> {
    Ok(dev.status())
}

// Non-blocking cache read, same shape as get_device_state — lets the UI recover which
// device is connected after reopening the cut dialog, without a redundant reconnect.
#[tauri::command]
pub fn get_connected_device(dev: tauri::State<DeviceManagerHandle>) -> Result<Option<DeviceInfo>, IpcError> {
    Ok(dev.connected.lock().unwrap().clone())
}

#[tauri::command]
pub fn plan_cut(state: tauri::State<AppStateHandle>, grouping: Grouping) -> Result<PlanCutResponse, IpcError> {
    plan_cut_response(&state.lock().unwrap().editor.doc, grouping)
}

#[tauri::command]
pub fn travel_for_order(
    state: tauri::State<AppStateHandle>,
    doc_revision: String,
    grouping: Grouping,
    passes: Vec<TravelPassDto>,
) -> Result<Vec<[f64; 4]>, IpcError> {
    // Fully qualified because the command and the function it forwards to share a name.
    crate::device::travel_for_order(&state.lock().unwrap().editor.doc, &doc_revision, grouping, &passes)
}

// async: prepare_cut briefly locks the document (plan + preflight), then the
// lock is dropped before execute_cut's blocking call into the device worker
// so `cut` never holds the doc mutex while blocked, and running off the main
// loop keeps the UI (and cancel_cut) responsive while it blocks.
#[tauri::command(async)]
pub fn cut(state: tauri::State<AppStateHandle>, dev: tauri::State<DeviceManagerHandle>, request: CutRequest) -> Result<CutStarted, IpcError> {
    let (planned_for, passes) = {
        let app = state.lock().unwrap();
        dev.prepare_cut(&app, request)?
    };
    dev.execute_cut(planned_for, passes)
}

#[tauri::command(async)]
pub fn cancel_cut(dev: tauri::State<DeviceManagerHandle>) -> Result<(), IpcError> {
    dev.cancel()
}

// async: blocks like `cut` while the worker drives the next pass.
#[tauri::command(async)]
pub fn resume_cut(dev: tauri::State<DeviceManagerHandle>) -> Result<(), IpcError> {
    dev.resume()
}

#[tauri::command(async)]
pub fn confirm_pass_done(dev: tauri::State<DeviceManagerHandle>) -> Result<(), IpcError> {
    dev.confirm_pass_done()
}

#[tauri::command]
pub fn list_presets(machine_id: String) -> Result<Vec<MaterialPreset>, IpcError> {
    crate::device::list_presets(&presets_path()?, &machine_id)
}

#[tauri::command]
pub fn machine_caps(dev: tauri::State<DeviceManagerHandle>, machine_id: String) -> Result<MachineCaps, IpcError> {
    dev.caps_for(&machine_id)
}

#[tauri::command]
pub fn save_preset(p: MaterialPreset) -> Result<(), IpcError> {
    crate::device::save_preset(&presets_path()?, p)
}

/// Keyed on the machine as well as the id: a preset id is the operator's own string, so the same
/// id can name a Cameo's material and a Puma's, and a delete that named only the id removed both
/// (#153).
#[tauri::command]
pub fn delete_preset(machine_id: String, id: String) -> Result<(), IpcError> {
    crate::device::delete_preset(&presets_path()?, &machine_id, &id)
}

fn presets_path() -> Result<PathBuf, IpcError> {
    cutplan::presets::default_presets_path()
        .ok_or_else(|| IpcError::new("no_config_dir", "cannot resolve presets file location"))
}

// async: reads each paired host's connection in the same order `list_devices` dials them, so
// against a sweep of unreachable hosts it trails one host behind and can approach the sweep's
// total — on the main thread that is a frozen window, not just a slow one.
#[tauri::command(async)]
pub fn list_hosts(dev: tauri::State<DeviceManagerHandle>) -> Result<Vec<PairedHostView>, IpcError> {
    Ok(dev.host_views())
}

/// Learn the fingerprint a host presents, so the pairing dialog can show it for confirmation
/// before anything is pinned. The first step of a first pairing, ahead of `test_host`: until the
/// operator has confirmed it there is no fingerprint to hand `test_host`.
///
/// async: the probe dials the host the same way `test_host` does (see there for why).
#[tauri::command(async)]
pub fn probe_host(address: String) -> Result<String, IpcError> {
    cut_host::client::probe_fingerprint(&address, cut_host::client::CONNECT_TIMEOUT)
        .map_err(|e| crate::device::host_error(&e))
}

/// Whether this address is already paired, and whether the certificate just probed is the one
/// pinned for it. Asked between the probe and the confirm, so a re-pairing after a Pi's
/// certificate changed is something the operator is told about rather than something they
/// discover as a second, permanently broken row (#107).
#[tauri::command]
pub fn existing_pairing(
    dev: tauri::State<DeviceManagerHandle>,
    address: String,
    fingerprint: String,
) -> Result<Option<ExistingPairing>, IpcError> {
    Ok(dev.existing_pairing(&address, &fingerprint))
}

/// Prove a host without saving it. The pairing dialog calls this before `pair_host`, so an
/// entry that has never worked is never written.
///
/// async: `pair_check` dials the host over the network (5s connect timeout with the DNS/mDNS
/// resolve bounded inside it, 30s body timeout) — on the main thread a mistyped address would
/// freeze the window for the length of that wait.
#[tauri::command(async)]
pub fn test_host(address: String, token: String, fingerprint: String) -> Result<Vec<DeviceInfo>, IpcError> {
    cut_host::client::HostClient::pair_check(&address, &token, &fingerprint)
        .map_err(|e| crate::device::host_error(&e))
}

// async: pairing dials the host the same way `test_host` does (see there for why).
#[tauri::command(async)]
pub fn pair_host(
    dev: tauri::State<DeviceManagerHandle>,
    name: String,
    address: String,
    token: String,
    fingerprint: String,
) -> Result<PairedHostView, IpcError> {
    let path = hosts_path()?;
    let id = dev.pair(name.clone(), address.clone(), token, fingerprint, &path)?;
    Ok(PairedHostView { id, name, address, unreachable: None })
}

// async: the idle check dials the host the same way `test_host` does (see there for why).
//
// `force` is the operator accepting that a host which cannot be asked may still be cutting. The
// UI only offers it after an unforced attempt has already been refused — see `DeviceManagerHandle::forget`.
#[tauri::command(async)]
pub fn forget_host(dev: tauri::State<DeviceManagerHandle>, id: HostId, force: bool) -> Result<(), IpcError> {
    dev.forget(&id, &hosts_path()?, force)
}

fn hosts_path() -> Result<PathBuf, IpcError> {
    crate::hosts::default_hosts_path()
        .ok_or_else(|| IpcError::new("no_config_dir", "this system has no configuration directory"))
}

/// Paths the user has actually chosen in the native picker this session.
///
/// The trace commands take a path from the webview, so without this any code running there could
/// name an arbitrary file and read image content back out. The set is only ever added to by
/// `pick_image`, which is the picker itself — the renderer cannot authorize a path, only ask for
/// one to be authorized by the user choosing it.
#[derive(Default)]
pub struct AuthorizedImages(pub Mutex<std::collections::HashSet<PathBuf>>);

/// Canonicalized so an authorized file cannot be re-reached under a different spelling
/// (`..` segments, a symlink, a relative path) and miss the membership check.
fn canonical(path: &PathBuf) -> Result<PathBuf, String> {
    std::fs::canonicalize(path).map_err(|e| format!("cannot read {}: {e}", path.display()))
}

/// Returns the *resolved* path to read, so the caller opens exactly what was authorized.
///
/// Returning `()` and letting the caller re-open its own argument would leave a window between
/// the check and the read: the caller's path is resolved twice, and a symlink that pointed at an
/// authorized file for the check can be retargeted before the open. Reading the already-resolved
/// path removes that step entirely.
fn authorized_path(auth: &AuthorizedImages, path: &PathBuf) -> Result<PathBuf, String> {
    let real = canonical(path)?;
    if auth.0.lock().unwrap().contains(&real) {
        Ok(real)
    } else {
        Err("that image was not selected in this session".into())
    }
}

/// Open the native image picker and record what the user chose. This lives in Rust rather than
/// in the webview precisely so that selection, not the caller's say-so, is what grants access.
#[tauri::command(async)]
pub fn pick_image(
    app: tauri::AppHandle,
    auth: tauri::State<AuthorizedImages>,
) -> Result<Option<PathBuf>, String> {
    use tauri_plugin_dialog::DialogExt;
    let picked = app
        .dialog()
        .file()
        .add_filter("Images", &["png", "jpg", "jpeg", "gif", "bmp"])
        .blocking_pick_file();
    let Some(picked) = picked else { return Ok(None) };
    let path = picked
        .into_path()
        .map_err(|e| format!("could not resolve the selected file: {e}"))?;
    let real = canonical(&path)?;
    auth.0.lock().unwrap().insert(real.clone());
    Ok(Some(real))
}

/// The tracer's own description of what it accepts, so the dialog renders its controls from the
/// module that enforces them rather than from a table typed to agree.
#[tauri::command]
pub fn trace_controls() -> Result<trace::TraceControlSpecs, IpcError> {
    Ok(trace::control_specs())
}

// async: load_system_fonts walks the system font directories on disk — keep the scan off the
// main thread, same reasoning as the trace commands. An empty list is honest data (a box with
// no fonts), not an error; the dialog renders it as a state.
#[tauri::command(async)]
pub fn list_fonts() -> Result<Vec<String>, IpcError> {
    Ok(geometry::list_font_families())
}

#[tauri::command(async)]
pub fn trace_image(
    auth: tauri::State<AuthorizedImages>,
    path: PathBuf,
    controls: trace::TraceControls,
) -> Result<trace::TraceResult, IpcError> {
    let real = authorized_path(&auth, &path).map_err(|m| IpcError::new("input", m))?;
    let bytes = trace::read_image(&real).map_err(trace_error)?;
    trace::trace(&bytes, &controls).map_err(trace_error)
}

/// Carry the tracer's own code across IPC, so the dialog branches on the kind of failure instead
/// of matching the text of one.
fn trace_error(e: trace::TraceError) -> IpcError {
    IpcError::new(e.code(), e.to_string())
}

/// Returns the source thumbnail as a re-encoded PNG data URL — never the file's original bytes,
/// so a path that is not a decodable image yields an error rather than its contents.
#[tauri::command(async)]
pub fn load_image_preview(
    auth: tauri::State<AuthorizedImages>,
    path: PathBuf,
) -> Result<String, String> {
    let real = authorized_path(&auth, &path)?;
    let bytes = trace::read_image(&real).map_err(|e| e.to_string())?;
    let png = trace::preview_png(&bytes).map_err(|e| e.to_string())?;
    use base64::Engine as _;
    Ok(format!("data:image/png;base64,{}", base64::engine::general_purpose::STANDARD.encode(png)))
}
